ocaml/typing/typecore.ml

```diff
@@ -6429,6 +6431,25 @@ and type_expect_
            exp_type = instance ty_expected;
            exp_attributes = sexp.pexp_attributes;
            exp_env = env }
+  | Pexp_stack e ->
+      let exp = type_expect env expected_mode e ty_expected_explained in
+      begin match exp.exp_desc with
+      | Texp_function { alloc_mode; _} | Texp_tuple (_, alloc_mode)
+      | Texp_construct (_, _, _, Some alloc_mode)
+      | Texp_variant (_, Some (_, alloc_mode))
+      | Texp_record {alloc_mode = Some alloc_mode; _}
+      | Texp_array (_, _, _, alloc_mode)
+      | Texp_field (_, _, _, Boxing (alloc_mode, _)) ->
+        begin match Locality.submode Locality.local
+          (Alloc.proj (Comonadic Areality) alloc_mode) with
+        | Ok () -> ()
+        | Error _ -> raise (Error (exp.exp_loc, env, Cannot_stack_allocate))
+        end
+      | _ ->
```

Comment:
A bunch of these cases *are* allocations -- they're just not allocations that can go on the stack. For example, `new cl` or `lazy`. Might be better to separate out allocations from non-allocations in the error message, so as not to confuse users.

Also, what about first-class modules? Can they be stack-allocated?

Reply:
Yes good point. Two solutions:
- For those (`new cl` or `lazy`) we will print a different message along the line of "stack-allocating this langauge construct is not supported yet"
- Or we rephrase `Not_allocation` to "This language construct cannot be stack-allocated", to cover both `new cl` and genuine non-allocation. 

I prefer the second one for simplicity. What do you think?

Reply:
first-class modules are similar to `new cl` and `lazy` - they are always heap-allocated.

Reply:
How about “This expression cannot be a stack-allocation site” ?

Reply:
Please treat the cases where we do not (yet?) support stack allocation separately from the expressions that are not allocations, and give them specific error messages too, for example:
```
Error: Objects cannot be allocated on the stack
```
You just need a variant type that represents all the things that can't be put on the stack, and it's much clearer to the user.